query.go

```diff
@@ -1,106 +1,203 @@
 package kallax
 
 import (
-	"encoding/json"
+	"fmt"
 
-	"github.com/src-d/go-kallax/operators"
-
-	"gopkg.in/mgo.v2/bson"
+	"github.com/Masterminds/squirrel"
 )
 
+// Query returns information about some query settings and compiles the query.
 type Query interface {
-	GetCriteria() bson.M
-	Sort(s Sort)
-	Limit(l int)
-	Skip(s int)
-	Select(p Select)
-	GetSort() Sort
-	GetLimit() int
-	GetSkip() int
-	GetSelect() Select
+	compile() ([]string, squirrel.SelectBuilder)
+	// GetOffset returns the number of skipped rows in the query.
+	GetOffset() uint64
+	// GetLimit returns the max number of rows retrieved by the query.
+	GetLimit() uint64
+	// GetBatchSize returns the number of rows retrieved by the store per
+	// batch. This is only used and has effect on queries with 1:N
+	// relationships.
+	GetBatchSize() uint64
+}
+
+type columnSet []string
+
+func (cs columnSet) contains(col string) bool {
+	for _, c := range cs {
+		if c == col {
+			return true
+		}
+	}
+	return false
+}
+
+func (cs *columnSet) add(cols ...string) {
+	for _, col := range cols {
+		cs.addCol(col)
+	}
+}
+
+func (cs *columnSet) addCol(col string) {
+	if !cs.contains(col) {
+		*cs = append(*cs, col)
+	}
 }
 
+func (cs columnSet) copy() []string {
+	var result = make(columnSet, len(cs))
+	for i, col := range cs {
+		result[i] = col
+	}
+	return result
+}
+
+// BaseQuery is a generic query builder.
 type BaseQuery struct {
-	clauses     []bson.M
-	limit, skip int
-	sort        Sort
-	selector    Select
-}
-
-func NewBaseQuery() *BaseQuery {
-	return &BaseQuery{clauses: make([]bson.M, 0)}
-}
-
-// AddCriteria adds a new mathing expression to the query, all the expressions
-// are merged on a $and expression.
-//
-// Use operators package instead of build expresion by hand:
-//
-//  import . "github.com/src-d/go-kallax/operators"
-//
-//  func (q *YourQuery) FindNonZeroRecords() {
-//      // All the Fields are defined on the Schema generated variable
-//      size := NewField("size", "int")
-//      q.AddCriteria(Gt(size, 0))
-//  }
-func (q *BaseQuery) AddCriteria(expr bson.M) {
-	q.clauses = append(q.clauses, expr)
-}
-
-// GetCriteria returns a valid bson.M used internally by Store.
-func (q *BaseQuery) GetCriteria() bson.M {
-	if len(q.clauses) == 0 {
-		return nil
+	columns         columnSet
+	excludedColumns columnSet
+	builder         squirrel.SelectBuilder
+
+	batchSize uint64
+	offset    uint64
+	limit     uint64
+}
+
+var _ Query = (*BaseQuery)(nil)
+
+// NewBaseQuery creates a new BaseQuery for querying the given table.
+func NewBaseQuery(table string) *BaseQuery {
+	return &BaseQuery{
+		builder: squirrel.StatementBuilder.
+			PlaceholderFormat(squirrel.Dollar).
+			Select().
+			From(table),
+		batchSize: 50,
 	}
+}
 
-	return operators.And(q.clauses...)
+// Select adds the given columns to the list of selected columns in the query.
+func (q *BaseQuery) Select(columns ...string) {
+	q.columns.add(columns...)
```

Comment:
Shouldn't `Select` remove the columns from the excluded set as well?

Think of:

```go
query.SelectNot('name')
// In other place
query.Select('name')
```

User should be able to select columns even if those were excluded before.